File: yao.jl/VQA/vqa.jl
#!/usr/bin/env julia

# Activate the local Yao.jl environment if available
try
    using Pkg
    project_root = joinpath(@__DIR__, "..")
    if isfile(joinpath(project_root, "Project.toml"))
        Pkg.activate(project_root)
        Pkg.instantiate()
    end
catch e
    @warn "Unable to activate Yao.jl project environment" error=e
end

using JSON
using LinearAlgebra
using Printf
using Random

# Load only the Yao components we need (avoids YaoPlots circular dependency issues)
using YaoBlocks
using YaoArrayRegister
import YaoBlocks: chain, control, put, Ry, X

const NUM_QUBITS = 4
const LAYERS = 3

const PAULI_MATRICES = Dict(
    'I' => Matrix{ComplexF64}(I, 2, 2),
    'X' => ComplexF64[0 1; 1 0],
    'Y' => ComplexF64[0 -im; im 0],
    'Z' => ComplexF64[1 0; 0 -1],
)

struct FrameworkPerformanceRow
    name::String
    energy::Float64
    iterations::Int
    runtime_ms::Float64
end

"""
Generate the STO-3G hydrogen Hamiltonian coefficients (Jordan–Wigner mapped).
Returns a vector of (label, coefficient) pairs where label is a 4-character Pauli string.
"""
function create_h2_hamiltonian_terms()
    terms = Vector{Tuple{String, Float64}}()
    push!(terms, ("IIII", -0.810_547_980_537_324))

    for (idx, coeff) in enumerate([
        0.172_183_932_619_155,
        0.172_183_932_619_155,
        -0.225_753_492_224_023,
        -0.225_753_492_224_023,
    ])
        label = ["I", "I", "I", "I"]
        label[idx] = "Z"
        push!(terms, (join(label), coeff))
    end

    for ((q1, q2), coeff) in [
        ((1, 2), 0.120_912_632_617_766),
        ((1, 3), 0.168_927_538_700_879),
        ((1, 4), 0.045_232_799_946_057),
        ((2, 3), 0.045_232_799_946_057),
        ((2, 4), 0.168_927_538_700_879),
        ((3, 4), 0.120_912_632_617_766),
    ]
        label = ["I", "I", "I", "I"]
        label[q1] = "Z"
        label[q2] = "Z"
        push!(terms, (join(label), coeff))
    end

    for ((q1, q2), coeff) in [
        ((1, 2), 0.166_145_432_563_824),
        ((3, 4), 0.174_643_430_683_004),
    ]
        for axis in ("X", "Y")
            label = ["I", "I", "I", "I"]
            label[q1] = axis
            label[q2] = axis
            push!(terms, (join(label), coeff))
        end
    end

    return terms
end

function pauli_matrix(label::String)
    mat = PAULI_MATRICES[label[1]]
    for idx in 2:length(label)
        mat = kron(mat, PAULI_MATRICES[label[idx]])
    end
    return mat
end

function build_hamiltonian_matrix(terms)
    dim = 2^NUM_QUBITS
    matrix = zeros(ComplexF64, dim, dim)
    for (label, coeff) in terms
        matrix .+= coeff * pauli_matrix(label)
    end
    return matrix
end

function compute_exact_ground_state_energy(matrix::Matrix{ComplexF64})
    eigenvalues = eigvals(Hermitian(matrix))
    return minimum(real.(eigenvalues))
end

function hardware_efficient_ansatz(params::AbstractVector{<:Real})
    expected_length = NUM_QUBITS * LAYERS
    @assert length(params) == expected_length "Expected $expected_length params, got $(length(params))"

    ops = Any[]
    idx = 1
    for _ in 1:LAYERS
        for wire in 1:NUM_QUBITS
            push!(ops, put(NUM_QUBITS, wire => Ry(params[idx])))
            idx += 1
        end
        for (control_wire, target_wire) in zip(1:NUM_QUBITS-1, 2:NUM_QUBITS)
            push!(ops, control(control_wire, target_wire => X))
        end
    end
    return chain(NUM_QUBITS, ops...)
end

function statevector(params::AbstractVector{<:Real})
    circuit = hardware_efficient_ansatz(params)
    reg = zero_state(NUM_QUBITS)
    reg = apply!(reg, circuit)
    return vec(reg.state)
end

function expectation_energy(params::AbstractVector{<:Real}, h_matrix::Matrix{ComplexF64})
    ψ = statevector(params)
    return real(dot(ψ, h_matrix * ψ))
end

function parameter_shift_gradient(cost_fn::Function, params::Vector{Float64})
    shift = π / 2
    grad = similar(params)
    temp = copy(params)
    for i in eachindex(params)
        temp[i] = params[i] + shift
        forward = cost_fn(temp)
        temp[i] = params[i] - shift
        backward = cost_fn(temp)
        grad[i] = 0.5 * (forward - backward)
        temp[i] = params[i]
    end
    return grad
end

function run_yao_vqe(h_matrix::Matrix{ComplexF64}, exact_energy::Float64)
    Random.seed!(1337)
    param_count = NUM_QUBITS * LAYERS
    params = 2π .* rand(param_count)
    cost_fn = θ -> expectation_energy(θ, h_matrix)

    lr = 0.01
    β1, β2 = 0.9, 0.999
    ε = 1e-8
    max_iters = 350
    tolerance = 1e-6

    m = zeros(param_count)
    v = zeros(param_count)
    iter = 0
    converged = false
    start_time = time()

    for t in 1:max_iters
        grad = parameter_shift_gradient(cost_fn, params)
        m = β1 .* m .+ (1 - β1) .* grad
        v = β2 .* v .+ (1 - β2) .* (grad .^ 2)
        m_hat = m ./ (1 - β1^t)
        v_hat = v ./ (1 - β2^t)
        params .-= lr .* m_hat ./ (sqrt.(v_hat) .+ ε)
        energy = cost_fn(params)
        iter = t

        if norm(grad) < tolerance
            converged = true
            break
        end
    end

    runtime_ms = (time() - start_time) * 1000
    final_energy = cost_fn(params)
    delta = abs(final_energy - exact_energy)

    return Dict(
        "framework" => "Yao.jl (Julia)",
        "exact_energy" => exact_energy,
        "vqe_energy" => final_energy,
        "energy_error" => delta,
        "iterations" => iter,
        "runtime_ms" => round(runtime_ms, digits=2),
        "parameters" => param_count,
        "converged" => converged,
    )
end

function print_framework_comparison(rows::Vector{FrameworkPerformanceRow}, exact_energy::Float64)
    println("\n" * "="^70)
    println("Cross-Framework VQE Comparison (H₂, STO-3G)")
    println("="^70)
    println(@sprintf("%-22s | %13s | %10s | %13s", "Framework", "Energy (Ha)", "Iterations", "Runtime (ms)"))
    println("-"^70)
    for row in rows
        println(@sprintf("%-22s | %13.6f | %10d | %13.2f", row.name, row.energy, row.iterations, row.runtime_ms))
    end
    best_row = rows[1]
    for row in rows[2:end]
        if row.energy < best_row.energy
            best_row = row
        end
    end
    println(@sprintf(
        "\nBest energy: %s (%.6f Ha, Δ vs exact = %.6f Ha)",
        best_row.name,
        best_row.energy,
        abs(best_row.energy - exact_energy),
    ))
end

function main()
    terms = create_h2_hamiltonian_terms()
    h_matrix = build_hamiltonian_matrix(terms)
    exact_energy = compute_exact_ground_state_energy(h_matrix)
    result = run_yao_vqe(h_matrix, exact_energy)
    
    # Write to JSON file
    output_file = get(ENV, "VQA_OUTPUT_FILE", "yao_vqa_result.json")
    open(output_file, "w") do f
        JSON.print(f, result, 2)
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: qiskit/VQA/vqa.py
"""
Cross-framework Variational Quantum Eigensolver (VQE) benchmark for H₂ using Qiskit.

This script mirrors the LogosQ Rust workflow by:
1. Building the STO-3G Hamiltonian (Jordan–Wigner mapped) with verified coefficients.
2. Computing the exact ground-state energy via dense diagonalization.
3. Running a hardware-efficient VQE (RealAmplitudes, linear entanglement, 3 reps) optimized
   with COBYLA to obtain a comparable variational energy.
4. Printing a compact benchmark table against reference PennyLane and LogosQ results.
"""

from __future__ import annotations

import json
import math
import sys
import time
from dataclasses import dataclass
from typing import List, Sequence, Tuple

import numpy as np
from qiskit import QuantumCircuit
from qiskit.circuit import Parameter
from qiskit.quantum_info import SparsePauliOp, Statevector
from scipy.optimize import minimize


@dataclass
class FrameworkPerformanceRow:
    name: str
    energy: float
    iterations: int
    runtime_ms: float


def _label_from_mutations(mutations: Sequence[Tuple[int, str]], num_qubits: int = 4) -> str:
    """Create a Pauli label where `mutations` supplies (qubit, axis) overrides."""
    label = ["I"] * num_qubits
    for qubit, axis in mutations:
        label[qubit] = axis
    return "".join(label)


def create_h2_hamiltonian() -> SparsePauliOp:
    """
    STO-3G Hamiltonian for H₂ at 0.735 Å using Jordan–Wigner mapping.
    Coefficients follow the canonical integrals widely cited in Qiskit tutorials.
    """
    terms: List[Tuple[str, float]] = []

    terms.append(("IIII", -0.810_547_980_537_324))

    for qubit, coeff in [
        (0, 0.172_183_932_619_155),
        (1, 0.172_183_932_619_155),
        (2, -0.225_753_492_224_023),
        (3, -0.225_753_492_224_023),
    ]:
        terms.append((_label_from_mutations([(qubit, "Z")]), coeff))

    for (q1, q2), coeff in [
        ((0, 1), 0.120_912_632_617_766),
        ((0, 2), 0.168_927_538_700_879),
        ((0, 3), 0.045_232_799_946_057),
        ((1, 2), 0.045_232_799_946_057),
        ((1, 3), 0.168_927_538_700_879),
        ((2, 3), 0.120_912_632_617_766),
    ]:
        terms.append((_label_from_mutations([(q1, "Z"), (q2, "Z")]), coeff))

    for (q1, q2), coeff in [
        ((0, 1), 0.166_145_432_563_824),
        ((2, 3), 0.174_643_430_683_004),
    ]:
        for axis in ("X", "Y"):
            terms.append((_label_from_mutations([(q1, axis), (q2, axis)]), coeff))

    return SparsePauliOp.from_list(terms)


def compute_exact_ground_state_energy(hamiltonian: SparsePauliOp) -> float:
    """Diagonalize the dense Hamiltonian matrix to obtain the exact ground energy."""
    matrix = hamiltonian.to_matrix()
    eigenvalues = np.linalg.eigvalsh(matrix)
    return float(np.min(eigenvalues).real)




def build_hardware_efficient_ansatz(num_qubits: int, reps: int = 3):
    """Build hardware-efficient ansatz: RY rotations + linear CNOT entanglement."""
    circuit = QuantumCircuit(num_qubits)
    params = []
    
    for layer in range(reps):
        layer_params = []
        for qubit in range(num_qubits):
            param = Parameter(f"θ_{layer}_{qubit}")
            layer_params.append(param)
            circuit.ry(param, qubit)
        params.extend(layer_params)
        
        # Linear CNOT entanglement
        for control in range(num_qubits - 1):
            circuit.cx(control, control + 1)
    
    return circuit, params


def run_qiskit_vqe(hamiltonian: SparsePauliOp, exact_energy: float, reps: int = 3) -> dict:
    """Execute a VQE with a hardware-efficient ansatz and Adam optimizer."""
    num_qubits = hamiltonian.num_qubits
    ansatz, param_list = build_hardware_efficient_ansatz(num_qubits, reps=reps)
    parameter_count = len(param_list)  # Should be reps * num_qubits
    h_matrix = hamiltonian.to_matrix()
    
    # Store parameter order for evaluation
    param_order = list(ansatz.parameters)

    def objective(theta: Sequence[float]) -> float:
        param_dict = {param: val for param, val in zip(param_order, theta)}
        bound = ansatz.assign_parameters(param_dict, inplace=False)
        state = Statevector.from_instruction(bound)
        amplitudes = state.data
        return float(np.real(np.vdot(amplitudes, h_matrix @ amplitudes)))

    # Adam optimizer implementation
    rng = np.random.default_rng(seed=1337)
    initial_point = 2 * math.pi * rng.random(size=parameter_count)
    
    # Adam optimizer parameters
    lr = 0.01
    beta1, beta2 = 0.9, 0.999
    epsilon = 1e-8
    max_iterations = 350
    tolerance = 1e-7
    
    # Adam state
    m = np.zeros(parameter_count)
    v = np.zeros(parameter_count)
    params = initial_point.copy()
    
    start = time.perf_counter()
    iterations = 0
    converged = False
    
    for t in range(1, max_iterations + 1):
        # Compute gradient using parameter-shift rule (for RY gates: shift = π/2)
        grad = np.zeros(parameter_count)
        shift = math.pi / 2
        for i in range(parameter_count):
            params_plus = params.copy()
            params_plus[i] += shift
            params_minus = params.copy()
            params_minus[i] -= shift
            grad[i] = 0.5 * (objective(params_plus) - objective(params_minus))
        
        # Adam update
        m = beta1 * m + (1 - beta1) * grad
        v = beta2 * v + (1 - beta2) * (grad ** 2)
        m_hat = m / (1 - beta1 ** t)
        v_hat = v / (1 - beta2 ** t)
        params = params - lr * m_hat / (np.sqrt(v_hat) + epsilon)
        
        energy = objective(params)
        iterations = t
        
        # Check convergence
        if np.linalg.norm(grad) < tolerance:
            converged = True
            break
    
    optimal_energy = objective(params)
    runtime_ms = (time.perf_counter() - start) * 1e3

    return {
        "framework": "Qiskit (Python)",
        "exact_energy": float(exact_energy),
        "vqe_energy": optimal_energy,
        "energy_error": abs(optimal_energy - exact_energy),
        "iterations": iterations,
        "runtime_ms": round(runtime_ms, 2),
        "parameters": parameter_count,
        "converged": converged,
    }


def print_framework_comparison(rows: Sequence[FrameworkPerformanceRow], exact_energy: float) -> None:
    print("\n" + "=" * 70)
    print("Cross-Framework VQE Comparison (H₂, STO-3G)")
    print("=" * 70)
    print(f"{'Framework':<22} | {'Energy (Ha)':>13} | {'Iterations':>10} | {'Runtime (ms)':>13}")
    print("-" * 70)
    for row in rows:
        print(f"{row.name:<22} | {row.energy:>13.6f} | {row.iterations:>10} | {row.runtime_ms:>13.2f}")

    best = min(rows, key=lambda r: r.energy)
    print(
        f"\nBest energy: {best.name} ({best.energy:.6f} Ha, Δ vs exact = {abs(best.energy - exact_energy):.6f} Ha)"
    )


def main() -> None:
    import os
    hamiltonian = create_h2_hamiltonian()
    exact_energy = compute_exact_ground_state_energy(hamiltonian)
    
    # Get number of layers from environment variable (default: 3 for 12 parameters)
    reps = int(os.environ.get("VQA_LAYERS", "3"))
    
    result = run_qiskit_vqe(hamiltonian, exact_energy, reps=reps)
    
    # Write to JSON file
    output_file = os.environ.get("VQA_OUTPUT_FILE", "qiskit_vqa_result.json")
    with open(output_file, "w") as f:
        json.dump(result, f, indent=2)


if __name__ == "__main__":
    main()
